### PR TITLE
Debuggability

### PR DESCRIPTION
**Pull request by youtux**
Improve "debuggability". Internal pytest-factoryboy calls are now visible when using a debugger like PDB or PyCharm.

TODO:

- [x] use different file name for the module created if the file already exist
- [x] assert that the make_module function does not override existing files, as that would be a bug. It actually happens if we register 2 factories having the same name. Add test coverage for this.
- [x] maybe the module path should contain the name of the module/class that defines it, so that we can have factories override at the class level?
- [x] try to use the same directory to create the generated files, so that we can debug them across invocations. A cache directory?
- [x] add `__init__.py` to the package created?

### CODE REVIEW

File: pytest_factoryboy/fixture.py [hunk 1]
```diff
@@ -1,51 +1,103 @@
...
 """
+)
 
 
-def make_fixture(name, module, func, args=None, related=None, **kwargs):
-    """Make fixture function and inject arguments.
+@dataclass
+class FixtureDef:
+    name: str
+    function_name: typing.Literal["model_fixture", "attr_fixture", "factory_fixture", "subfactory_fixture"]
```
Comment (skarzi) [inline pytest_factoryboy/fixture.py:72]: What do you think about using `enum.Enum` to define all possible function names?

File: pytest_factoryboy/fixture.py [hunk 2]
```diff
@@ -1,51 +1,103 @@
...
+from pytest_factoryboy.fixture import (
+    attr_fixture,
+    factory_fixture,
+    model_fixture,
+    subfactory_fixture,
+)
+
+__all__ = (
+% for fixture_def in fixture_defs:
+    ${ repr(fixture_def.name)},
```
Comment (skarzi) [inline pytest_factoryboy/fixture.py:40]:
```suggestion
    ${ repr(fixture_def.name) },
```

File: pytest_factoryboy/fixture.py [hunk 3]
```diff
@@ -1,51 +1,103 @@
...
+def _fixture(related):
+    def fixture_maker(fn):
+        fn._factoryboy_related = related
+        return pytest.fixture(fn)
+
+    return fixture_maker
+% for fixture_def in fixture_defs:
+
+
+${ fixture_def.kwargs_var_name } = {}
```
Comment (skarzi) [inline pytest_factoryboy/fixture.py:54]: Why setting pytest-factoryboy fixtures kwargs has 2-steps?
Couldn't we simply render `fixture_def.function_kwargs` in fixture definition (in line 63)?

File: pytest_factoryboy/fixture.py [hunk 4]
```diff
@@ -1,51 +1,103 @@
...
+        return f"_{self.name}__kwargs"
 
-    if related:
-        fixture_func._factoryboy_related = related
 
-    fixture = pytest.fixture(fixture_func)
-    setattr(module, name, fixture)
-    return fixture
+@lru_cache()  # This way we reuse the same folder for the whole execution of the program
+def make_temp_folder(package_name):
```
Comment (skarzi) [inline pytest_factoryboy/fixture.py:83]:
```suggestion
def make_temp_folder(package_name: str) -> Path:
```

File: pytest_factoryboy/fixture.py [hunk 5]
```diff
@@ -1,51 +1,103 @@
...
+    """Create a temporary folder and automatically delete it when the process exit."""
+    path = pathlib.Path(tempfile.mkdtemp()) / package_name
+    path.mkdir(parents=True, exist_ok=True)
+
+    atexit.register(shutil.rmtree, str(path))
+
+    return path
+
+
+def make_module(code: str, module_name: str, package_name: str):
```
Comment (skarzi) [inline pytest_factoryboy/fixture.py:93]: NIP: let's add missing type hinting for return type:

```suggestion
def make_module(code: str, module_name: str, package_name: str) -> ModuleType:
```

File: pytest_factoryboy/fixture.py [hunk 6]
```diff
@@ -99,41 +154,55 @@ def register(factory_class, _name=None, **kwargs):
...
+            name=model_name,
+            function_name="model_fixture",
+            function_kwargs={"factory_name": factory_name},
+            deps=deps,
+            related=related,
+        )
     )
+
+    code = module_template.render(fixture_defs=fixture_defs)
+    mod = make_module(code, module_name=model_name, package_name="_pytest_factoryboy_generated_fixtures")
```
Comment (skarzi) [inline pytest_factoryboy/fixture.py:198]: NIP: what do you think about using some more meaningful name e.g. `generated_module` or `model_fixtures_module` etc?

File: pytest_factoryboy/fixture.py [hunk 7]
```diff
@@ -1,51 +1,103 @@
...
+
+    return fixture_maker
+% for fixture_def in fixture_defs:
+
+
+${ fixture_def.kwargs_var_name } = {}
+
+
+@_fixture(related=${ repr(fixture_def.related) })
+def ${ fixture_def.name }(
```
Comment (skarzi) [inline pytest_factoryboy/fixture.py:58]: NIP: now, we could even add simple docstrings to fixtures. Something like:

+ `my_app.MyModel instances factory.`
+ `Factory of MyModelFactory.my_attribute attribute.`

Docstrings generation can be implemented on `FixtureDef` dataclass.

**Comment by youtux** [comment]
Hi @skarzi, thank you for your valuable review comments. I'm currently on holiday, but I plan to get back to it somewhere next week.

File: pytest_factoryboy/fixture.py (hunk 1 above)
Comment (youtux) [reply pytest_factoryboy/fixture.py:89]: We wouldn't reference that enum anywhere else, so I don't find it useful for now.

File: pytest_factoryboy/fixture.py (hunk 7 above)
Comment (youtux) [reply pytest_factoryboy/fixture.py:66]: that could be a good idea, but I don't feel like thinking about the docstring at the moment, maybe an improvement for a next iteration

File: pytest_factoryboy/fixture.py (hunk 3 above)
Comment (youtux) [reply pytest_factoryboy/fixture.py:62]: function_kwargs contain objects that are not just strings, but they can be any object.
Stringifying those objects will just not work. We need to inject them after the module is executed

**Pull request merged**